File: lexers/OCaml/ocaml-examples-3.08/basics/sieve_vect.c
/*(***********************************************************************)
(*                                                                     *)
(*                           Objective Caml                            *)
(*                                                                     *)
(*               Pierre Weis, projet Cristal, INRIA Rocquencourt       *)
(*                                                                     *)
(*  Copyright 2001 Institut National de Recherche en Informatique et   *)
(*  en Automatique.  All rights reserved.  This file is distributed    *)
(*  under the terms of the Q Public License version 1.0.               *)
(*                                                                     *)
(***********************************************************************)

(* Erathostene sieve, imperative version.
   A vector is initialized with integers.
   Then the vector is sieved. *)

*/

#include <stdio.h>

#define fixed_bound 5000000

long v[fixed_bound];

void sieve (long max) {
 long i;
 long j;
 long prime_count = 0;
 long prime;

 for (i = 0; i < max; i++) {
   v[i] = i + 1;
 };

 v[0] = 0;
 prime_count = 0;

 for (i = 0; i < max; i++) {
   if (v[i] != 0) {
     prime_count = prime_count + 1;
     prime = i + 1;
     for (j = i + prime; j < max; j = j + prime) {
       if (j < max) {
         v[j] = 0;
       };
     };
   };
 };
 /* for (i = 0; i < max; i++) {printf ("%d ", v[i]);};*/
 printf ("There are %d primes less than or equal to %d.\n", prime_count, max);
}

void main () {
  sieve (fixed_bound);
}

File: lexers/OCaml/ocaml-examples-3.08/basics/sieve_vect.ml
(***********************************************************************)
(*                                                                     *)
(*                           Objective Caml                            *)
(*                                                                     *)
(*               Pierre Weis, projet Cristal, INRIA Rocquencourt       *)
(*                                                                     *)
(*  Copyright 2001 Institut National de Recherche en Informatique et   *)
(*  en Automatique.  All rights reserved.  This file is distributed    *)
(*  only by permission.                                                *)
(*                                                                     *)
(***********************************************************************)

(* Erathostene sieve, imperative version.
   A vector is initialized with consecutive integers.
   Then the vector is sieved by removing multiples of the next still
   not removed integer, starting from 2. *)

let fixed_bound = 5000000;;

let sieve max =

 let v = Array.init max (fun i -> i + 1) in
 let prime_count = ref 0 in

 v.(0) <- 0;
 prime_count := 0;

 for i = 1 to max - 1 do
  if v.(i) <> 0 then begin
   prime_count := !prime_count + 1;
   let prime = i + 1 in
   let rec sieve j =
    if j < max then begin v.(j) <- 0; sieve (j + prime) end in
   sieve (i + prime)
  end
 done;
 Printf.printf
  "There are %d primes less than or equal to %d.\n" !prime_count max;;

let main () =
 let max =
   try int_of_string (Sys.argv.(1))
   with _ -> fixed_bound in
 sieve max;;

main ();;
